### PR TITLE
Step 49 modernization

### PR DESCRIPTION
**Pull request by drwells**
This is the little project that I have been working on that prompted the changes to CylindricalManifold: I redid the discussion to use the Manifold classes and, while I was around, added some C++11 features to the rest of the tutorial.

### CODE REVIEW

**Comment by drwells** [comment]
I also have updates: these should be `step-49.yuhan.2.png` through `step-49.yuhan.5.png`.

grid-0-b
grid-1-b
grid-2-b
grid-3-b

**Comment by drwells** [comment]
Part of #4800.

**Comment by drwells** [comment]
One problem with this is that we don't version tutorial images on the website: if we update these then we will update the example images for 8.5 and older.

We could simply include them in the repo directly: they are only about 100k each and I can compress them down more if we need to.

**Comment by bangerth** [comment]
We can version these files if you just provide a different name in `results.dox` where you reference them. Or we could just not worry about the issue -- are the resulting meshes substantially different?

**Comment by drwells** [comment]
The images are fairly different since the current set show what happens if one uses `CylinderBoundary` and the new ones show what happens for `CylindricalManifold`.

Versioning is the right idea: the new images should be named `step-49.yuhan.6.png` through `step-49.yuhan.9.png` (I don't have a better idea for a naming convention).

**Comment by bangerth** [comment]
OK. If you send me a tar file with the pictures (and correct filenames) you want put on the web server, then I'll put them there.

**Comment by drwells** [comment]
Done

**Comment by drwells** [comment]
@bangerth I mixed up the renumbering. Let me try this again.

**Comment by drwells** [comment]
The numbers should now be correct, both in the emailed tarball and in the HTML document.

I got confused because the new version has only four pictures, while the old one had six.

**Comment by bangerth** [comment]
The updated figures are now online.

I'm booked for the rest of the day. Can someone else read through the patch and if appropriate merge?

**Comment by drwells** [comment]
I can confirm that the images are up and now show up nicely in my local copy.

File: examples/step-49/doc/results.dox [hunk 1]
```diff
@@ -225,158 +239,109 @@ void create_3d_grid (Triangulation<3> &triangulation)
...
-
-The next step is to teach each of the top surfaces that they should be
-curved. We can do this by creating CylinderBoundary objects that
-describe this. A first attempt looks like this:
+This creates the following mesh:
+
+<img src="https://www.dealii.org/images/steps/developer/step-49.yuhan.8.png"
+     alt="" width="400" height="355">
+
+This mesh the right general shape, but the top cells are now polygonal: their
```
Comment (masterleinad) [inline examples/step-49/doc/results.dox]: This mesh HAS the

File: examples/step-49/doc/results.dox [hunk 2]
```diff
@@ -225,158 +239,109 @@ void create_3d_grid (Triangulation<3> &triangulation)
...
+and associating it with the cells above $y = 3$. This way, when we refine the
+cells on top, we will place new points along concentric circles instead of
+straight lines.
+
+In deal.II we describe all geometries with classes that inherit from
+Manifold. The default geometry is Cartesian and is implemented in the
+FlatManifold class. As the name suggests, Manifold and its inheriting classes
+provide a way to describe curves and curved cells in a general way with ideas
+and terminology from differential geometry: for example, CylindricalManifold
+inherits from ChartManifold, which is describes a geometry through pull backs
```
Comment (masterleinad) [inline examples/step-49/doc/results.dox]: ~~is~~

File: examples/step-49/doc/results.dox [hunk 3]
```diff
@@ -225,158 +239,109 @@ void create_3d_grid (Triangulation<3> &triangulation)
...
+geodesics are arcs along circles orthogonal to the $z$-axis centered along the
+line $(0, 3, z)$.
+
+Since all three top parts of the domain use the same geodesics, we will
+mark all cells with centers above the $y = 3$ line as being cylindrical in
+nature:
 
 @code
+  const Tensor<1, 3> axis {{0.0, 0.0, 1.0}};
+  const Point<3> axial_point {0, 3.0, 0.0};
```
Comment (masterleinad) [inline examples/step-49/doc/results.dox]: The different number of curly brackets looks inconsistent.

File: examples/step-49/doc/results.dox [hunk 4]
```diff
@@ -225,158 +239,109 @@ void create_3d_grid (Triangulation<3> &triangulation)
...
-  static const CylinderBoundary<3> inner_cylinder (inner_radius,
-                                                   Point<3>(0,0,1),
-                                                   Point<3>(0,3,0));
-  static const CylinderBoundary<3> outer_cylinder (outer_radius,
-                                                   Point<3>(0,0,1),
-                                                   Point<3>(0,3,0));
-  triangulation.set_boundary (9, outer_cylinder);
-@endcode
+  const Tensor<1, 3> axis {{0.0, 0.0, 1.0}};
+  const Point<3> axial_point {0, 3.0, 0.0};
```
Comment (masterleinad) [inline examples/step-49/doc/results.dox]: The different number of curly brackets looks inconsistent. Isn't a single pair sufficient?

File: examples/step-49/doc/results.dox (hunk 1 above)
Comment (drwells) [reply examples/step-49/doc/results.dox]: fixed.

File: examples/step-49/doc/results.dox (hunk 2 above)
Comment (drwells) [reply examples/step-49/doc/results.dox]: fixed.

File: examples/step-49/doc/results.dox (hunk 3 above)
Comment (drwells) [reply examples/step-49/doc/results.dox]: You're right. Unfortunately, in C++11 we need two pairs for the Tensor. I changed the Point constructor to be a function call.

I don't know how else to get two `const` objects here.

**Comment by bangerth** [comment]
That's because the `Tensor` constructor takes an array, whereas `Point<3>` takes three separate arguments.

**Comment by drwells** [comment]
Hm. I suppose we are stuck with the current version, then.

**Comment by drwells** [comment]
I don't have a strong preference here. Would you prefer
```cpp
Tensor<1, 3> axis({0.0, 0.0, 1.0});
```
to the current version
```cpp
Tensor<1, 3> axis {{0.0, 0.0, 1.0}};
```
I can change it.

**Comment by jppelteret** [comment]
Very nice! Thanks for taking the time to make this update :-)

**Comment by masterleinad** [comment]
Yes, I would prefer
```
const Tensor<1, 3> axis ({0.0, 0.0, 1.0});
const Point<3> axial_point(0, 3.0, 0.0);
```
to
```
const Tensor<1, 3> axis {{0.0, 0.0, 1.0}};
const Point<3> axial_point{0, 3.0, 0.0};
```

**Comment by drwells** [comment]
Easy enough. Fixed :)

**Comment by masterleinad** [comment]
Thanks!

**Comment by drwells** [comment]
@jppelteret Thanks. We always need more Manifold tutorials :)

**Pull request merged**